chain/store/name.go

```diff
@@ -0,0 +1,276 @@
+		if height, ok := key.(uint32); ok {
+			if nc, ok := value.(namesCleanup); ok && len(nc) > 0 {
+				sdb.updateNamesCleanup(height, nc)
+			} else {
+				sdb.deleteNamesCleanup(height)
+			}
+			if commit {
+				sdb.namesCleanup.Delete(height)
+			}
+		}
+		return true
+	})
+}
+
+func (sdb *StateDB) getNamesCleanup(height uint32) (namesCleanup, error) {
+	if v, ok := sdb.namesCleanup.Load(height); ok {
+		if nc, ok := v.(namesCleanup); ok {
+			return nc, nil
+		}
+	}
+
+	enc, err := sdb.trie.TryGet(append(NameCleanupPrefix, getNamesCleanupId(height)...))
+	if err != nil {
+		return nil, fmt.Errorf("[getNamesCleanup]can not get name cleanup from trie: %v", err)
+	}
+
+	nc := make(namesCleanup, 0)
+
+	if len(enc) > 0 {
+		buff := bytes.NewBuffer(enc)
+		ncLength, err := serialization.ReadVarUint(buff, 0)
+		if err != nil {
+			return nil, fmt.Errorf("[getNamesCleanup]Failed to decode state object for name cleanup: %v", err)
+		}
+		for i := uint64(0); i < ncLength; i++ {
+			id, err := serialization.ReadVarString(buff)
+			if err != nil {
+				return nil, fmt.Errorf("[getNamesCleanup]Failed to decode state object for name cleanup: %v", err)
+			}
+			nc[id] = struct{}{}
+		}
+	}
+
+	sdb.namesCleanup.Store(height, nc)
+	return nc, nil
+}
+
+func getNamesCleanupId(height uint32) []byte {
+	buff := bytes.NewBuffer(nil)
+	serialization.WriteUint32(buff, height)
+	return buff.Bytes()
+}
+
+func (sdb *StateDB) CleanupNames(height uint32) error {
+	ids, err := sdb.getNamesCleanup(height)
+	if err != nil {
+		return err
+	}
+	for id := range ids {
+		sdb.names.Store(id, nil)
+	}
+	sdb.namesCleanup.Store(height, nil)
+
+	return nil
+}
+
+func (sdb *StateDB) cleanupNamesAtHeight(height uint32, id []byte) error {
+	ids, err := sdb.getNamesCleanup(height)
+	if err != nil {
+		return err
+	}
+	ids[string(id)] = struct{}{}
+	return nil
+}
+
+func (sdb *StateDB) deleteNamesCleanup(height uint32) error {
+	err := sdb.trie.TryDelete(append(NameCleanupPrefix, getNamesCleanupId(height)...))
+	if err != nil {
+		return err
+	}
+
+	sdb.namesCleanup.Delete(height)
+	return nil
+}
+
+func (sdb *StateDB) updateNamesCleanup(height uint32, nc namesCleanup) error {
+	buff := bytes.NewBuffer(nil)
+
+	if err := serialization.WriteVarUint(buff, uint64(len(nc))); err != nil {
+		panic(fmt.Errorf("can't encode names cleanup %v: %v", nc, err))
+	}
+	ncs := make([]string, 0, len(nc))
+	for id := range nc {
+		ncs = append(ncs, id)
+	}
+	sort.Strings(ncs)
+	for _, id := range ncs {
+		if err := serialization.WriteVarString(buff, id); err != nil {
+			panic(fmt.Errorf("can't encode names cleanup %v: %v", nc, err))
+		}
+	}
+
+	return sdb.trie.TryUpdate(append(NameCleanupPrefix, getNamesCleanupId(height)...), buff.Bytes())
+}
+
+func (sdb *StateDB) cancelNameCleanupAtHeight(height uint32, name string) error {
+	ids, err := sdb.getPubSubCleanup(height)
+	if err != nil {
+		return err
+	}
+	if _, ok := ids[name]; ok {
+		delete(ids, name)
+	}
```

Comment:
It is unnecessary for `comma ok` before `delete`. Just `delete(ids, name)` is enough.

[Reference](https://golang.org/src/builtin/builtin.go?h=delete#L146)
```
// The delete built-in function deletes the element with the specified key
// (m[key]) from the map. If m is nil or there is no such element, delete
// is a no-op.
func delete(m map[Type]Type1, key Type)
```